vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -135,25 +141,34 @@ def refresh_metrics_metadata_cache(self):
         # Apparently only when the server restarts?
         # https://pubs.vmware.com/vsphere-50/index.jsp?topic=%2Fcom.vmware.wssdk.pg.doc_50%2FPG_Ch16_Performance.18.5.html
 
-    def refresh_tags_cache(self):
-        # type: () -> None
+    def collect_tags(self, infrastructure_data):
+        # type: (InfrastructureData) -> ResourceTags
         """
         Fetch the all tags, build tags for each monitored resources and store all of that into the tags_cache.
         """
         if not self.api_rest:
-            return
-        mors_iterator = itertools.chain.from_iterable(
-            self.infrastructure_cache.get_mors(resource_type) for resource_type in self.config.collected_resource_types
-        )
+            return {}
+
+        # In order to be more efficient in tag collection, the infrastructure data is filtered as much as possible.
+        # All filters are applied except the ones based on tags of course.
+        resource_filters_without_tags = [f for f in self.config.resource_filters if not isinstance(f, TagFilter)]
+        filtered_infra_data = {
+            mor: props
+            for mor, props in iteritems(infrastructure_data)
+            if isinstance(mor, tuple(self.config.collected_resource_types))
```

Comment:
Maybe we can filter the infra data upfront (just after `infrastructure_data = self.api.get_infrastructure()`)  to avoid having this condition twice: `isinstance(mor, tuple(self.config.collected_resource_types))`

Reply:
I gave some details in the PR description:

> Why applying filters twice? Two reasons:
> - Maintainability: It is easier to make sure that filtering is correct if we only apply it once to A.
> - Separation of concerns: the initial filtering from A to B is only for improved performance of tags collection.

Reply:
> Maintainability: It is easier to make sure that filtering is correct if we only apply it once to A.

I'm not sure to understand why it's more maintainable to filter twice. I'm referring to this duplication:
- https://github.com/DataDog/integrations-core/blob/e2e23e62c97f35cd7b2824b80e360a2ea37c136d/vsphere/datadog_checks/vsphere/vsphere.py#L196-L198
- https://github.com/DataDog/integrations-core/blob/e2e23e62c97f35cd7b2824b80e360a2ea37c136d/vsphere/datadog_checks/vsphere/vsphere.py#L158

That can be done just once `refresh_infrastructure_cache` and we can pass the filtered result with only the right resource types mors to `self.collect_tags(...)`.

> Separation of concerns: the initial filtering from A to B is only for improved performance of tags collection.

Filtering the infra data with the right types and then using it later doesn't seems to be a breach of separation of concerns.

---

That said, I don't feel strongly about either way. I'm fine if you prefer keep it that way :)

Reply:
Oh sorry, I thought your comment included both the `isinstance` check and the application of filters 🤦 

Yeah that makes sense 👍

Reply:
Ok I tried to address that with this commit: https://github.com/DataDog/integrations-core/pull/6297/commits/8385109bea42c302704ee9932606c34f5f3f3dc3

I have mixed feeling, as it requires to keep in memory both the filtered and the original data.
The reason is some features require access to the complete infrastructure tree in order to get information about the parents of each resource recursively. (Those features being the addition of tags coming from the parent resources, and the filter on 'inventory path').

Indeed even if a resource is ignored by filter, it will be used to create tags on its child resources. (example: even though you only monitor VMs, some information will be pulled from ESXi hosts in order to add tags to those VMs like `vsphere_host:<HOSTNAME>`)

I think I'd rather revert that commit, what do you think?

Reply:
Hum, but `filtered_infra_data` will be garbage collected once we exit `refresh_infrastructure_cache()` right?

Yeah, from readability standpoint, I think it's quite similar. I prefer the refactored version, but I'm ok with both, the different is minor. Up to you :)

Reply:
They are both gonna be garbage collected yes :)